src/date.c

```diff
@@ -1,18 +1,1006 @@
+  return msPerDay * ecma_DayFromYear(y);
+}
+
+static int ecma_YearFromTime_s(etime_t t)
+{
+  int first = floor((t / msPerDay) / 366) + 1970,
+              last = floor((t / msPerDay) / 365) + 1970, middle = 0;
+  
+  if (last < first) {
+    int temp = first;
+    first = last;
+    last = temp;
+  }
+  
+  while (last > first) {
+    middle = (last + first) / 2;
+    if (ecma_TimeFromYear(middle) > t) {
+      last = middle - 1;
+    } else {
+      if (ecma_TimeFromYear(middle) <= t) {
+        if (ecma_TimeFromYear(middle + 1) > t) {
+          first = middle;
+          break;
+        }
+        first = middle + 1;
+      }
+    }
+  }
+  
+  return first;
+}
+
+static int ecma_InLeapYear(etime_t t, int year) {
+  (void)t;
+  return ecma_DaysInYear(year) == 366;
+}
+
+static int ecma_DayWithinYear(etime_t t, int year) {
+  return (int)(ecma_Day(t) - ecma_DayFromYear(year));
+}
+
+static void ecma_getfirstdays(int* days, int isleap) {
+  unsigned int i;
+  static int sdays[] = {0, 31, 59, 90, 120, 151, 181, 212,
+                        243, 273, 304, 334, 365};
+  memcpy(days, sdays, sizeof(sdays));
+  
+  if(isleap) {
+    for(i = 2; i < ARRAY_SIZE(sdays); i++) {
+      days[i] += 1;
+    }
+  }
+}
+
+static int ecma_MonthFromTime(etime_t t, int year) {
+  int days[MonthInYear+1];
+  etimeint_t dwy = ecma_DayWithinYear(t, year);
+  int ily = ecma_InLeapYear(t, year);
+  
+  unsigned int i; int ret = -1;
+  
+  ecma_getfirstdays(days, ily);
+  
+  for(i = 0; i < ARRAY_SIZE(days) - 1; i++) {
+    if(dwy >= days[i] && dwy < days[i+1]) {
+      ret = i;
+      break;
+    }
+  }
+  
+  return ret;
+}
+
+static int ecma_DateFromTime(etime_t t, int year) {
+  int days[MonthInYear+1];
+  int mft = ecma_MonthFromTime(t, year);
+  int dwy = ecma_DayWithinYear(t, year);
+  int ily = ecma_InLeapYear(t, year);
+  
+  if(mft > 11) {
+    return -1;
+  }
+  
+  ecma_getfirstdays(days, ily);
+  
+  return dwy - days[mft] + 1;
+}
+
+static int ecma_WeekDay(etime_t t) {
+  return (ecma_Day(t)+4) % 7;
+}
+
+static int ecma_DaylightSavingTA(etime_t t) {
+  time_t time = t / 1000;
+  struct tm tm;
+  memset(&tm, 0, sizeof(t));
+  localtime_r(&time, &tm);
+  if(tm.tm_isdst > 0) {
+    return msPerHour;
+  } else {
+    return 0;
+  }
+}
+
+static int ecma_LocalTZA() {
+  return (int)-timezone * 1000;
+}
+
+static etimeint_t ecma_LocalTime(etime_t t) {
+  return t + ecma_LocalTZA() + ecma_DaylightSavingTA(t);
+}
+
+static etimeint_t ecma_UTC(etime_t t) {
+  return t - ecma_LocalTZA() - ecma_DaylightSavingTA(t-ecma_LocalTZA());
+}
+
+static int ecma_HourFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerHour) % HoursPerDay;
+}
+
+static int ecma_MinFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerMinute) % MinutesPerHour;
+}
+
+static int ecma_SecFromTime(etime_t t) {
+  return (etimeint_t)floor(t / msPerSecond) % SecondsPerMinute;
+}
+
+static int ecma_msFromTime(etime_t t) {
+  return (etimeint_t)t % msPerSecond;
```

Comment:
I was confused a bit since according to 
http://www.urbandictionary.com/define.php?term=Dito

Dito
Used commonly by Puerto Ricans or Dominicans as a word to express ptty..like saying "aww poor you"

=))